### PR TITLE
Added option to fetch and upload old PI file to from and to S3, and added logic to update old PI file

### LINKED ISSUE #27: Fetch and update old pi file to s3
Required by #21 and #22 

- Need to generate an initial old pi file containing pi names and month.
- Upload this initial pi file to B2 in a known location manually.
- Update the code to fetch the old pi file.
- Update the code to write a new pi file with the new pis encountered this month.
- Upload the old (new) pi file to s3.

### PR DESCRIPTION
**Pull request by QuanMPhm**
Partially closes #27. If the user does not provide an old_pi file, `process_report.py` will by default fetch the file from a hardcoded location in S3 storage.
Setting the `upload-to-s3` option will now upload both the processed CSV invoices and the updated PI file.

I've also renamed some functions for consistent capitalization.

I have also added a few print statements to provide the following info:
- When a new PI is encountered in an invoice
- When a future PI is found in a past invoice

It should be noted that if future PIs are encountered when processing old invoices, the invoice script will consider them new PIs for the purpose of the new PI credit, and will update the future PI's first-month date to be same as the old invoice's date.

Blocked on #23, and prerequisite to #21

### CODE REVIEW

File: process_report/process_report.py [hunk 1]
```diff
@@ -58,7 +58,14 @@ def load_old_pis(old_pi_file):
...
 
 
+def dump_old_pis(old_pi_file, old_pi_dict: dict):
+    with open(old_pi_file, "w") as f:
+        for pi, first_month in old_pi_dict.items():
+            f.write(f"{pi},{first_month}\n")
+
+
 def is_old_pi(old_pi_dict, pi, invoice_month):
+    # If we're processing old invoices, this will flag PIs who were new *back then* as old. But this would only be the case if future PIs appeared in past invoices???
```
Comment (knikolla) [inline process_report/process_report.py:68]: - Do not write lines this long, break the comment into multiple line comments.
- Comments are not a place to have a dialogue with yourself. An issue with the code should be flagged as an issue, or resolved.
- You can easily turn this check from a `!=` to comparing if the date is in the past or future. 
  - Write a unit test for it.

File: process_report/process_report.py [hunk 2]
```diff
@@ -217,9 +228,10 @@ def main():
             invoice_list.append(os.path.join(args.output_folder, pi_invoice))
 
         upload_to_s3(invoice_list, invoice_month)
+        upload_to_s3_old_pi_file(old_pi_file)
 
 
-def fetch_S3_invoices(invoice_month):
+def fetch_s3_invoices(invoice_month):
```
Comment (knikolla) [inline process_report/process_report.py:234]: While yes, this should have been `s3` instead of `S3` and in python things are preferred to be written in lower case snake case. Please try to refrain from introducing unrelated changes to a PR. Do not revert it for now, this is fine.

**Review by knikolla — CHANGES_REQUESTED** [review]

**Comment by QuanMPhm** [comment]
I've added some small features and have updated the PR summary

File: process_report/process_report.py [hunk 3]
```diff
@@ -58,8 +58,16 @@ def load_old_pis(old_pi_file):
...
+def dump_old_pis(old_pi_file, old_pi_dict: dict):
+    with open(old_pi_file, "w") as f:
+        for pi, first_month in old_pi_dict.items():
+            f.write(f"{pi},{first_month}\n")
+
+
 def is_old_pi(old_pi_dict, pi, invoice_month):
-    if pi in old_pi_dict and old_pi_dict[pi] != invoice_month:
+    if pi in old_pi_dict and old_pi_dict[pi] > invoice_month:
+        print(f"Found future PI {pi} from {old_pi_dict[pi]} in {invoice_month} invoice")
```
Comment (knikolla) [inline process_report/process_report.py:69]: This print is not necessary and this function can be simplified into only one `if` statement by using `.get` on the dictionary and providing a default value of the current month.

**Review by knikolla — CHANGES_REQUESTED** [review]

File: process_report/process_report.py (hunk 3 above)
Comment (QuanMPhm) [reply process_report/process_report.py:69]: We don't need to know if a future PI is encountered?

Comment (knikolla) [reply process_report/process_report.py]: There's no such thing as a "future" PI. If when re-processing the invoices of the past, you encounter a PI whose first encountered invoice is later than the invoice you're processing, there's something wrong with the invoices or the PI file. If you want to be extra careful you can raise an exception here and just stop the script altogether with an error.

Comment (QuanMPhm) [reply process_report/process_report.py]: I've made it raise an exception, and added a test case for it.

File: process_report/process_report.py [hunk 4]
```diff
@@ -58,8 +58,16 @@ def load_old_pis(old_pi_file):
...
 
+def dump_old_pis(old_pi_file, old_pi_dict: dict):
+    with open(old_pi_file, "w") as f:
+        for pi, first_month in old_pi_dict.items():
+            f.write(f"{pi},{first_month}\n")
+
+
 def is_old_pi(old_pi_dict, pi, invoice_month):
-    if pi in old_pi_dict and old_pi_dict[pi] != invoice_month:
+    if pi in old_pi_dict and old_pi_dict[pi] > invoice_month:
```
Comment (knikolla) [inline process_report/process_report.py:68]: Rewrite this block as below to improve readability. 
```python
first_invoice_month = old_pi_dict.get(pi, invoice_month)
if first_invoice_month > invoice_month:
    ... exit with the error here
if first_invoice_month < invoice_month:
    return True
return False
```


```

**Review by knikolla — CHANGES_REQUESTED** [review]
One last comment, otherwise looks good.

**Review by knikolla — APPROVED** [review]
Looks good to me. @naved001 can you do a final pass today?

**Review by naved001 — CHANGES_REQUESTED** [review]

**Comment by QuanMPhm** [comment]
@knikolla @naved001 Which PR should be pushed first?

**Comment by naved001** [comment]
@QuanMPhm maybe the order you proposed the PRs. You'll need to fix the merge conflicts on both PRs though.
